### PR TITLE
Add Jewish calendar

### PR DESCRIPTION
**Pull request by chaimt**
Add support to add jewish birthdays

### CODE REVIEW

**Comment by pinkywafer** [comment]
Thanks for this!
I'll have to take a while to review this as I have no experience of the hebrew calendar!
Can you please also update the info and readme docs.

**Comment by chaimt** [comment]
updated documention

**Comment by pinkywafer** [comment]
ok, @chaimt a few points to go over:
* I'm unsure whether the calendar should be defined and return `Jewish` or `Hebrew`? Either way, I think I'd rather have continuity - At the moment, it is configured as `jewish`, the `calendar_type` attribute returns `jewish`, but the date is listed as `hebrew_date` rather than `jewish_date`.  Could these ALL be either `hebrew` or `jewish` please.
* since the `hebrew_date` attribute is only populated when configured as jewish calendar type, please only return it if configured.
* id_prefix should not be returned as an attribute - it is only used to configure the entity name
* There should not be `@property` methods for attributes, please remove the extra ones for calendar_type and hebrew_date
* templates do not work - in the `async_update` function, `validate_date` must have the calendar type passed to it to prevent it breaking, otherwise it assumes a gregorian date which fails for hebrew dates.
* you have (accidentally I assume!) removed the `icon_soon` line from the readme, please reinstate this
* please also update info.md
* Can you add to the config flow for configuration?
* Currently, any entities configured in config flow do not save the calendar_type, so that needs to be handled by the sensor, even if configuration is added for config_flow as pre-existing entities will not have this saved.

File: custom_components/anniversaries/sensor.py [hunk 1]
```diff
@@ -119,6 +182,11 @@ def device_state_attributes(self):
         except:
             res[ATTR_DATE] = self._date
         res[ATTR_WEEKS] = self._weeks_remaining
+        res[ATTR_CALENDAR_TYPE] = self._calendar_type
+        if self._calendar_type == HEBREW_CALENDAR:
+            res[ATTR_HEBREW_DATE] = "{year}-{month}-{day}".format(year=self._h_date.year,month=self._h_date.month,day=self._h_date.day)
+        else:
+            res[ATTR_HEBREW_DATE] = ""
```
Comment (pinkywafer) [inline custom_components/anniversaries/sensor.py:189]: Only return hebrew dates etc if configured:
```suggestion
        if self._calendar_type == HEBREW_CALENDAR:
            res[ATTR_CALENDAR_TYPE] = self._calendar_type
            res[ATTR_HEBREW_DATE] = "{year}-{month}-{day}".format(year=self._h_date.year,month=self._h_date.month,day=self._h_date.day)

```

File: custom_components/anniversaries/sensor.py [hunk 2]
```diff
@@ -143,7 +211,7 @@ async def async_update(self):
         if self._template_sensor:
             try:
                 template_date = templater.Template(self._date_template, self.hass).async_render()
-                self._date, self._unknown_year = validate_date(template_date)
+                self._date, self._unknown_year = validate_date(template_date, "")
```
Comment (pinkywafer) [inline custom_components/anniversaries/sensor.py:214]:
```suggestion
                self._date, self._unknown_year = validate_date(template_date, self._calendar_type)

```

File: info.md [hunk 3]
```diff
@@ -27,6 +27,8 @@ Attributes:
 * unit_of_measurement: 'Days' By default, this is displayed after the state. _this is NOT translate-able.  See below for work-around_
 * half_anniversary_date: The date of the next half anniversary (if enabled by `show_half_anniversary`)
 * days_until_half_anniversary: The number of days until the next half anniversary
+* calendar_type: Type of calendar to use for entry (gregorian, hebrew) **Default**: gregorian
```
Comment (pinkywafer) [inline info.md:30]:
```suggestion
* calendar_type: returns hebrew if configured using hebrew calendar
```

File: info.md [hunk 4]
```diff
@@ -27,6 +27,8 @@ Attributes:
...
+* hebrew_date: Will display original hebrew data configured
```
Comment (pinkywafer) [inline info.md:31]:
```suggestion
* hebrew_date: Will display original hebrew date if configured using hebrew date
```

File: README.md [hunk 5]
```diff
@@ -84,6 +84,10 @@ anniversaries:
 | `icon_today` | Yes | Icon if the anniversary is today **Default**: `mdi:calendar-star`
 | `days_as_soon` | Yes | Days in advance to display the icon defined in `icon_soon` **Default**: 1
 | `icon_soon` | Yes | Icon if the anniversary is 'soon' **Default**: `mdi:calendar`
+| `calendar_type` | Yes | Type of calendar to use for entry (gregorian, hebrew) **Default**: gregorian
+| `hebrew_date` | Yes | will display original hebrew data configured
+
+
```
Comment (pinkywafer) [inline README.md:90]: this configuration option has no effect
```suggestion
```

File: README.md [hunk 6]
```diff
@@ -100,11 +104,34 @@ anniversaries:
...
 
 Unit_of_measurement is *not* translate-able.
 You can, however, change the text for unit of measurement in the configuration.  NB the sensor will always report in days, this just allows you to represent this in your own language.
 
+### Hebrew Dates
+[Hebrew-Calendar]: https://en.wikipedia.org/wiki/Hebrew_calendar
+This is a different a calendar then the gregorian calendar.
+The date you but in the by the hebrew date, and the calculation will be when it is this year on the gregorian calendar.
+The date field will be propagated with the gregorian date of this year and then date original date will be saved in a seperate filed (hebrew_date)
+
```
Comment (pinkywafer) [inline README.md:119]: remove extra blank lines
```suggestion
```

File: custom_components/anniversaries/const.py [hunk 7]
```diff
@@ -37,6 +37,8 @@
 CONF_ONE_TIME = "one_time"
 CONF_DATE_EXCLUSION_ERROR = "Configuration cannot include both `date` and `date_template`. configure ONLY ONE"
 CONF_DATE_REQD_ERROR = "Either `date` or `date_template` is Required"
+CONF_CALENDAR_TYPE = "calendar_type"
+CONF_HEBREW_DATE = "hebrew_date"
```
Comment (pinkywafer) [inline custom_components/anniversaries/const.py:41]: unneeded
```suggestion
```

File: custom_components/anniversaries/const.py [hunk 8]
```diff
@@ -86,6 +89,9 @@ def check_date(value):
         vol.Optional(CONF_UNIT_OF_MEASUREMENT, default=DEFAULT_UNIT_OF_MEASUREMENT): cv.string,
         vol.Optional(CONF_ID_PREFIX, default=DEFAULT_ID_PREFIX): cv.string,
         vol.Optional(CONF_ONE_TIME, default=DEFAULT_ONE_TIME): cv.boolean,
+        vol.Optional(CONF_CALENDAR_TYPE, default=DEFAULT_CALENDAR_TYPE): cv.string,
+        vol.Optional(CONF_HEBREW_DATE, default=""): cv.string,
```
Comment (pinkywafer) [inline custom_components/anniversaries/const.py:93]: unneeded
```suggestion
```

File: custom_components/anniversaries/const.py [hunk 9]
```diff
@@ -86,6 +89,9 @@ def check_date(value):
...
+
```
Comment (pinkywafer) [inline custom_components/anniversaries/const.py:94]: remove blank lines
```suggestion
```

File: custom_components/anniversaries/sensor.py [hunk 10]
```diff
@@ -33,25 +37,68 @@
...
+        try:
+            return datetime.strptime(value, "%m-%d"), True
+        except ValueError:
+                return "Invalid Date", False
+
+
+def is_not_blank(s):
+    return bool(s and s.strip())
+
+
```
Comment (pinkywafer) [inline custom_components/anniversaries/sensor.py:101]: this is not used as far as I can see
```suggestion
```

File: custom_components/anniversaries/sensor.py [hunk 11]
```diff
@@ -104,10 +155,22 @@ def state(self):
...
+        """Return the type of calendar."""
+        return self._calendar_type
+
+    def hebrew_date(self):
+        """Return the hebrew date."""
+        if self.calendar_type == HEBREW_CALENDAR:
+            return self._h_date
+        else:
+            return None
+
```
Comment (pinkywafer) [inline custom_components/anniversaries/sensor.py:168]: these are never used
```suggestion
```

File: info.md [hunk 12]
```diff
@@ -72,6 +74,8 @@ anniversaries:
 | `icon_today` | Yes | Icon if the anniversary is today **Default**: `mdi:calendar-star`
 | `days_as_soon` | Yes | Days in advance to display the icon defined in `icon_soon` **Default**: 1
 | `icon_soon` | Yes | Icon if the anniversary is 'soon' **Default**: `mdi:calendar`
+| `calendar_type` | Yes | Type of calendar to use for entry (gregorian, hebrew) **Default**: gregorian
+| `hebrew_date` | Yes | will display original hebrew data configured
```
Comment (pinkywafer) [inline info.md:78]: config not used
```suggestion
```

File: README.md [hunk 13]
```diff
@@ -100,11 +104,34 @@ anniversaries:
 * unit_of_measurement: 'Days' By default, this is displayed after the state. _this is NOT translate-able.  See below for work-around_
 * half_anniversary_date: The date of the next half anniversary (if enabled by `show_half_anniversary`)
 * days_until_half_anniversary: The number of days until the next half anniversary
-
+* hebrew_date: original hebrew date (if hebrew calendar)
+* calendar_type: type of calendar
```
Comment (pinkywafer) [inline README.md:108]:
```suggestion
* calendar_type: hebrew if configured using hebrew calendar

```

File: README.md [hunk 14]
```diff
@@ -84,6 +84,10 @@ anniversaries:
 | `icon_today` | Yes | Icon if the anniversary is today **Default**: `mdi:calendar-star`
 | `days_as_soon` | Yes | Days in advance to display the icon defined in `icon_soon` **Default**: 1
 | `icon_soon` | Yes | Icon if the anniversary is 'soon' **Default**: `mdi:calendar`
+| `calendar_type` | Yes | Type of calendar to use for entry (gregorian, hebrew) **Default**: gregorian
```
Comment (pinkywafer) [inline README.md:87]: currently the hebrew date is only available in yaml mode config, so:
```suggestion
| `calendar_type` | Yes | Type of calendar to use for entry (gregorian, hebrew) **Default**: gregorian _(Note this is ONLY available in YAML configuration)_
```

**Review by pinkywafer — CHANGES_REQUESTED** [review]
Please see the comments on the files
I think a rebase will also be needed as there have been changes to the repo since this PR was opened

**Pull request closed without merging**